src/OrleansGoogleUtils/Providers/Streams/GooglePubSubAdapterReceiver.cs

```diff
@@ -0,0 +1,150 @@
+            string deploymentId, IGooglePubSubDataAdapter dataAdapter, TimeSpan? deadline = null)
+        {
+            if (queueId == null) throw new ArgumentNullException(nameof(queueId));
+            if (dataAdapter == null) throw new ArgumentNullException(nameof(dataAdapter));
+            if (serializationManager == null) throw new ArgumentNullException(nameof(serializationManager));
+
+            var pubSub = new GooglePubSubDataManager(projectId, topicId, queueId.ToString(), deploymentId, deadline);
+            return new GooglePubSubAdapterReceiver(serializationManager, queueId, pubSub, dataAdapter);
+        }
+
+        private GooglePubSubAdapterReceiver(SerializationManager serializationManager, QueueId queueId, GooglePubSubDataManager pubSub, IGooglePubSubDataAdapter dataAdapter)
+        {
+            Id = queueId ?? throw new ArgumentNullException(nameof(queueId));
+            _serializationManager = serializationManager;
+            _pubSub = pubSub ?? throw new ArgumentNullException(nameof(pubSub));
+            _dataAdapter = dataAdapter ?? throw new ArgumentNullException(nameof(dataAdapter));
+            _logger = LogManager.GetLogger(GetType().Name, LoggerType.Provider);
+            _pending = new List<PendingDelivery>();
+        }
+
+        public Task Initialize(TimeSpan timeout)
+        {
+            if (_pubSub != null) return _pubSub.Initialize();
+
+            return Task.CompletedTask;
+        }
+
+        public async Task Shutdown(TimeSpan timeout)
+        {
+            try
+            {
+                // await the last pending operation, so after we shutdown and stop this receiver we don't get async operation completions from pending operations.
+                if (_outstandingTask != null)
+                    await _outstandingTask;
+            }
+            finally
+            {
+                // remember that we shut down so we never try to read from the queue again.
+                _pubSub = null;
+            }
+        }
+
+        public async Task<IList<IBatchContainer>> GetQueueMessagesAsync(int maxCount)
+        {
+            try
+            {
+                var pubSubRef = _pubSub; // store direct ref, in case we are somehow asked to shutdown while we are receiving.    
+                if (pubSubRef == null) return new List<IBatchContainer>();
+
+                int count = maxCount < 0 ? 10 : maxCount;
+
+                var task = pubSubRef.GetMessages(count);
+                _outstandingTask = task;
+                IEnumerable<ReceivedMessage> messages = await task;
+
+                List<IBatchContainer> pubSubMessages = new List<IBatchContainer>();
+                foreach (var message in messages)
+                {
+                    IBatchContainer container = _dataAdapter.FromPullResponseMessage(message.Message, _lastReadMessage++);
+                    pubSubMessages.Add(container);
```

Comment:
I read through google pubsub. Found out they have two modes for subsribers, push and pull. Each mode is there for fulfilling different set of use cases. https://cloud.google.com/pubsub/docs/subscriber

Just curious, have you considered enable push model for GooglePubSubStreamProvider?